packages/cdk/lib/stacks/context-stack.ts

```diff
@@ -1,31 +1,129 @@
 import { Stack, StackProps } from "monocdk";
-import { Vpc } from "monocdk/aws-ec2";
+import { IVpc, Vpc } from "monocdk/aws-ec2";
 import { Construct } from "constructs";
 import { getCommonParameter } from "../util";
 import { VPC_PARAMETER_NAME } from "../constants";
 import { ContextAppParameters } from "../env";
-import { BatchStack } from "./nested/batch-stack";
-import { CromwellEngineStack, CromwellEngineStackProps } from "./nested/cromwell-engine-stack";
-import { NextflowEngineStack, NextflowEngineStackProps } from "./nested/nextflow-engine-stack";
+import { BatchStack, BatchStackProps } from "./nested/batch-stack";
+import { CromwellEngineStack } from "./nested/cromwell-engine-stack";
+import { NextflowEngineStack } from "./nested/nextflow-engine-stack";
 import { ManagedPolicy } from "monocdk/aws-iam";
 
 export interface ContextStackProps extends StackProps {
   readonly contextParameters: ContextAppParameters;
 }
 
 export class ContextStack extends Stack {
+  private readonly vpc: IVpc;
+
   constructor(scope: Construct, id: string, props: ContextStackProps) {
     super(scope, id, props);
 
     const vpcId = getCommonParameter(this, VPC_PARAMETER_NAME);
-    const vpc = Vpc.fromLookup(this, "Vpc", { vpcId });
+    this.vpc = Vpc.fromLookup(this, "Vpc", { vpcId });
+
+    const { contextParameters } = props;
+    const { engineName } = contextParameters;
+
+    switch (engineName) {
+      case "cromwell":
+        this.renderCromwellStack(props);
+        break;
+      case "nextflow":
+        this.renderNextflowStack(props);
+        break;
+      default:
+        throw Error(`Engine '${engineName}' is not supported`);
+    }
+  }
+
+  private renderCromwellStack(props: ContextStackProps) {
+    const batchProps = this.getCromwellBatchProps(props);
+    const batchStack = this.renderBatchStack(batchProps);
+
+    let jobQueue;
+    if (props.contextParameters.requestSpotInstances) {
+      jobQueue = batchStack.batchSpot.jobQueue;
+    } else {
+      jobQueue = batchStack.batchOnDemand.jobQueue;
+    }
```

Comment:
This smells like a leaky abstraction. We've gotten to a state where `BatchStack` is trying to represent two different states based on the compute engine it's using, resulting in complexity across multiple classes to handle the different types of state.

Rather than make `BatchStack` fit every use case, what if we rendered batches on demand?  Then getting out the batch we want for Cromwell would be 

```
const batchStack = this.renderBatchStack(batchProps);
const computeType = props.contextParameters.requestSpotInstances ? ComputeType.SPOT : ComputeType.ON_DEMAND;
const jobQueue = batchStack.renderBatch("TaskBatch", computeType).jobQueue;
```

and Nextflow

```
const batchStack = this.renderBatchStack(batchProps);
let jobQueue, headQueue;
if (props.contextParameters.requestSpotInstances) {
  jobQueue = batchStack.renderBatch("TaskBatch", ComputeType.SPOT).jobQueue;
  headQueue = batchStack.renderBatch("TaskBatch", ComputeType.ON_DEMAND).jobQueue;
} else {
  headQueue = jobQueue = batchStack.renderBatch("TaskBatch", ComputeType.ON_DEMAND).jobQueue;
}
```